python-package/lightgbm/basic.py

```diff
@@ -5281,3 +5281,162 @@ def __get_eval_info(self) -> None:
                 self.__higher_better_inner_eval = [
                     name.startswith(("auc", "ndcg@", "map@", "average_precision")) for name in self.__name_inner_eval
                 ]
+
+
+class ObjectiveFunction:
+    """
+    ObjectiveFunction in LightGBM.
+
+    This class exposes the builtin objective functions for evaluating gradients and hessians
+    on external datasets. LightGBM does not use this wrapper during its training as it is
+    using the underlying C++ class.
+    """
```

Comment:
Maybe we should add some documentation on how this class should be used. Afaict, you cannot pass it as `objective` to `lgb.train` (would this be possible? 🤔) and you only use it to get access to the C++ implementation of `convert_outputs` and `get_gradients` if you have a need to do so, right?

Reply:
I will update the documentation! However I am against making it possible to use the `ObjectiveFunction` class as an objective function directly as that does not seem useful.